src/rulesets/LinearAlgebra/structured.jl

```diff
@@ -35,6 +35,16 @@ end
 ##### `Diagonal`
 #####
 
+_sqrt_pullback(Δ::Tangent, y) = return NoTangent(), Δ / (2y)
+_sqrt_pullback(Δ::Diagonal, y) = return NoTangent(), Δ / (2y)
+_sqrt_pullback(Δ::Matrix, y) = return NoTangent(), Diagonal(Δ) / (2y)
```

Comment:
Note that this allocates twice (or 3 times) when it need only do once:
```
julia> y = Diagonal(rand(10^3)); Δ = Diagonal(rand(10^3));

julia> @btime $Δ / (2*$y);
  458.116 ns (2 allocations: 15.88 KiB)

julia> @btime Diagonal($Δ.diag ./ (2 .* $y.diag));
  242.366 ns (1 allocation: 7.94 KiB)
```
I also think `_sqrt_pullback` ought to have a more specific name, since it's not the gradient for cases other than Diagonal. Or clearer, just methods of the closure `sqrt_pullback`, since you don't dispatch on `y`, and don't re-use these elsewhere.

Reply:
```
julia> m = rand(10^3, 10^3);

julia> @btime Diagonal($m) / (2 * $y);
  1.188 μs (3 allocations: 23.81 KiB)

julia> @btime Diagonal(view($m, LinearAlgebra.diagind($m)) ./ (2 .* $y.diag));
  531.579 ns (3 allocations: 8.02 KiB)
```

Reply:
If I replace the internal functions with methods of the closure, any suggestions for a more specific name if you still think that clarifies things? Something like `sqrt_on_diag_pullback`?

Also, just for convention sake, would if switching to methods of the closure rather than internal functions, would they stay outside the `rrule` or all be inside?

Reply:
I was thinking like this
```
function ChainRulesCore.rrule(::typeof(sqrt), d::Diagonal)
    y = sqrt(d)
    @assert y isa Diagonal
    sqrt_pullback(Δ::Union{Tangent, Diagonal}) = Diagonal(Δ.diag ./ (2 .* y.diag))
    sqrt_pullback(Δ::AbstractMatrix) = Diagonal(view(Δ, diagind(Δ)) ./ (2 .* y.diag))
    return y, sqrt_pullback
end
```
but then I don't see the tidy way to unthunk things.

Reply:
I think this makes sense to me. The idea being that unthunking is a cheap operation and can be done harmlessly on something already unthunked?

Reply:
Yes, it should compile away completely on anything else.

Reply:
Oh, my `Δ isa AbstractMatrix` case will catch Diagonal too, that's not ideal. Could define, for possible re-use in other rules, something like 
```
_diagview(x::Diagonal) = x.daig
_diagview(x::AbstractMatrix) = view(x, diagind(x))

_diagview(x::Tangent{<:Diagonal}) = x.diag # maybe
```
but Zygote IIRC makes a Tangent{Any}, so I don't know.

Reply:
Hmm, unless I have implemented this wrong, it appears to disagree with the tests

```
julia> include("test/runtests.jl")
Testing ChainRules.jl
Testing test_helpers.jl:
  0.336034 seconds (229.34 k allocations: 12.901 MiB, 34.21% gc time, 82.39% compilation time)

Testing rulesets/LinearAlgebra/structured.jl:
test_rrule: sqrt on Diagonal{Float64, Vector{Float64}}: Error During Test at /Users/thomaswright/.julia/packages/ChainRulesTestUtils/Rzheq/src/testers.jl:191
  Got exception outside of a @test

test_rrule: sqrt on Diagonal{Float64, Vector{Float64}}: Test Failed at /Users/thomaswright/.julia/packages/ChainRulesTestUtils/Rzheq/src/testers.jl:211
  Expression: ad_cotangents isa Tuple
  Problem: The pullback must return a Tuple (∂self, ∂args...)
   Evaluated: [0.11801667283102346 0.0 0.0; 0.0 0.12251226526582518 0.0; 0.0 0.0 0.09027072617315017] isa Tuple
Stacktrace:
  [1] #invokelatest#2
    @ ./essentials.jl:708 [inlined]
  [2] invokelatest(::Any)
    @ Base ./essentials.jl:706
  [3] evaluate_call_recurse!(recurse::Any, frame::Frame, call_expr::Expr; enter_generated::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:229
  [4] evaluate_call_recurse!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:201 [inlined]
  [5] eval_rhs(recurse::Any, frame::Frame, node::Expr)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:400
  [6] step_expr!(recurse::Any, frame::Frame, node::Any, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:561
  [7] step_expr!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:611
  [8] finish!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:14
  [9] finish_and_return!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:30 [inlined]
 [10] evaluate_call_recurse!(recurse::Any, frame::Frame, call_expr::Expr; enter_generated::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:240
--- the last 7 lines are repeated 5 more times ---
 [46] evaluate_call_recurse!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:201 [inlined]
 [47] eval_rhs(recurse::Any, frame::Frame, node::Expr)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:400
 [48] step_expr!(recurse::Any, frame::Frame, node::Any, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:561
 [49] step_expr!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:611
 [50] finish!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:14
 [51] finish_and_return!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:30 [inlined]
 [52] finish_and_return! (repeats 2 times)
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:34 [inlined]
 [53] top-level scope
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/construct.jl:781
 [54] include
    @ ./Base.jl:387 [inlined]
 [55] macro expansion
    @ ./timing.jl:210 [inlined]
 [56] include_test(path::String)
    @ Main ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:27
 [57] macro expansion
    @ ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:63 [inlined]
 [58] macro expansion
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
--- the last 2 lines are repeated 1 more time ---
 [61] macro expansion
    @ ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:59 [inlined]
 [62] macro expansion
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
test_rrule: sqrt on Diagonal{Float64, Vector{Float64}}: Test Failed at /Users/thomaswright/.julia/packages/ChainRulesTestUtils/Rzheq/src/testers.jl:215
  Expression: length(ad_cotangents) == length(primals)
  Problem: The pullback should return 1 cotangent for the primal and each primal input.
   Evaluated: 9 == 2
Stacktrace:
  [1] #invokelatest#2
    @ ./essentials.jl:708 [inlined]
  [2] invokelatest(::Any)
    @ Base ./essentials.jl:706
  [3] evaluate_call_recurse!(recurse::Any, frame::Frame, call_expr::Expr; enter_generated::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:229
  [4] evaluate_call_recurse!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:201 [inlined]
  [5] eval_rhs(recurse::Any, frame::Frame, node::Expr)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:400
  [6] step_expr!(recurse::Any, frame::Frame, node::Any, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:561
  [7] step_expr!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:611
  [8] finish!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:14
  [9] finish_and_return!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:30 [inlined]
 [10] evaluate_call_recurse!(recurse::Any, frame::Frame, call_expr::Expr; enter_generated::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:240
--- the last 7 lines are repeated 5 more times ---
 [46] evaluate_call_recurse!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:201 [inlined]
 [47] eval_rhs(recurse::Any, frame::Frame, node::Expr)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:400
 [48] step_expr!(recurse::Any, frame::Frame, node::Any, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:561
 [49] step_expr!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:611
 [50] finish!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:14
 [51] finish_and_return!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:30 [inlined]
 [52] finish_and_return! (repeats 2 times)
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:34 [inlined]
 [53] top-level scope
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/construct.jl:781
 [54] include
    @ ./Base.jl:387 [inlined]
 [55] macro expansion
    @ ./timing.jl:210 [inlined]
 [56] include_test(path::String)
    @ Main ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:27
 [57] macro expansion
    @ ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:63 [inlined]
 [58] macro expansion
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
--- the last 2 lines are repeated 1 more time ---
 [61] macro expansion
    @ ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:59 [inlined]
 [62] macro expansion
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
test_rrule: sqrt on Diagonal{Float64, Vector{Float64}}: Test Failed at /Users/thomaswright/.julia/packages/ChainRulesTestUtils/Rzheq/src/testers.jl:233
  Expression: ad_cotangent isa NoTangent
   Evaluated: 0.11801667283102346 isa NoTangent
Stacktrace:
  [1] #invokelatest#2
    @ ./essentials.jl:708 [inlined]
  [2] invokelatest(::Any)
    @ Base ./essentials.jl:706
  [3] evaluate_call_recurse!(recurse::Any, frame::Frame, call_expr::Expr; enter_generated::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:229
  [4] evaluate_call_recurse!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:201 [inlined]
  [5] eval_rhs(recurse::Any, frame::Frame, node::Expr)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:400
  [6] step_expr!(recurse::Any, frame::Frame, node::Any, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:561
  [7] step_expr!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:611
  [8] finish!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:14
  [9] finish_and_return!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:30 [inlined]
 [10] evaluate_call_recurse!(recurse::Any, frame::Frame, call_expr::Expr; enter_generated::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:240
--- the last 7 lines are repeated 5 more times ---
 [46] evaluate_call_recurse!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:201 [inlined]
 [47] eval_rhs(recurse::Any, frame::Frame, node::Expr)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:400
 [48] step_expr!(recurse::Any, frame::Frame, node::Any, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:561
 [49] step_expr!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:611
 [50] finish!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:14
 [51] finish_and_return!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:30 [inlined]
 [52] finish_and_return! (repeats 2 times)
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:34 [inlined]
 [53] top-level scope
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/construct.jl:781
 [54] include
    @ ./Base.jl:387 [inlined]
 [55] macro expansion
    @ ./timing.jl:210 [inlined]
 [56] include_test(path::String)
    @ Main ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:27
 [57] macro expansion
    @ ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:63 [inlined]
 [58] macro expansion
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
--- the last 2 lines are repeated 1 more time ---
 [61] macro expansion
    @ ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:59 [inlined]
 [62] macro expansion
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
test_rrule: sqrt on Diagonal{Float64, Vector{Float64}}: Error During Test at /Users/thomaswright/.julia/packages/ChainRulesTestUtils/Rzheq/src/check_result.jl:24
  Test threw exception
  Expression: isapprox(actual, expected; kwargs...)

test_rrule: sqrt on Diagonal{Float64, Vector{Float64}}: Error During Test at /Users/thomaswright/.julia/packages/ChainRulesTestUtils/Rzheq/src/testers.jl:191
  Got exception outside of a @test

  1.217141 seconds (3.60 M allocations: 296.746 MiB, 6.53% gc time, 27.01% compilation time)

Test Summary:                                                    | Pass  Fail  Error  Total
ChainRules                                                       |  641     3      3    647
  test_helpers.jl                                                |    8                   8
  rulesets                                                       |  633     3      3    639
    LinearAlgebra                                                |  633     3      3    639
      Structured Matrices                                        |  633     3      3    639
        / and \ on Square Matrixes                               |   81                  81
        Diagonal                                                 |   16                  16
        dot(x, ::Diagonal, y)                                    |   11                  11
        ::Diagonal * ::AbstractVector                            |    9                   9
        diag                                                     |   52                  52
        diagm                                                    |   24                  24
        sqrt(::Diagonal)                                         |    2     3      3      8
          test_rrule: sqrt on Diagonal{Float64, Vector{Float64}} |    1            1      2
          test_rrule: sqrt on Diagonal{Float64, Vector{Float64}} |    1     3      2      6
        Adjoint, Float64                                         |   30                  30
        Adjoint, ComplexF64                                      |   30                  30
        adjoint, Float64                                         |   30                  30
        adjoint, ComplexF64                                      |   30                  30
        Transpose, Float64                                       |   30                  30
        Transpose, ComplexF64                                    |   30                  30
        transpose, Float64                                       |   30                  30
        transpose, ComplexF64                                    |   30                  30
        UpperTriangular                                          |    7                   7
        LowerTriangular                                          |    7                   7
        triu                                                     |   47                  47
        tril                                                     |   47                  47
        det and logdet Diagonal                                  |   30                  30
        det and logdet UpperTriangular                           |   30                  30
        det and logdet LowerTriangular                           |   30                  30
ERROR: LoadError: Some tests did not pass: 641 passed, 3 failed, 3 errored, 0 broken.
in expression starting at /Users/thomaswright/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:34
```

Reply:
That is with this:

```
_diagview(x::Diagonal) = x.daig
_diagview(x::AbstractMatrix) = view(x, diagind(x))
_diagview(x::Tangent{<:Diagonal}) = x.diag
function ChainRulesCore.rrule(::typeof(sqrt), d::Diagonal)
    y = sqrt(d)
    @assert y isa Diagonal
    function sqrt_pullback(Δ_raw)
      Δ = unthunk(Δ_raw)
      Δ_diag = _diagview(Δ)
      return Diagonal(Δ_diag ./ (2 .* y.diag))
    end
    return y, sqrt_pullback
end
```

Reply:
I think this might be a `Int` vs `Float` thing again

Reply:
Changing the tests to contain exclusively floats fixed one of the three errors:

```
Testing ChainRules.jl
Testing test_helpers.jl:
  0.168716 seconds (229.38 k allocations: 12.902 MiB, 85.17% compilation time)

Testing rulesets/LinearAlgebra/structured.jl:
test_rrule: sqrt on Diagonal{Float64, Vector{Float64}}: Error During Test at /Users/thomaswright/.julia/packages/ChainRulesTestUtils/Rzheq/src/testers.jl:191
  Got exception outside of a @test

test_rrule: sqrt on Diagonal{Float64, Vector{Float64}}: Test Failed at /Users/thomaswright/.julia/packages/ChainRulesTestUtils/Rzheq/src/testers.jl:211
  Expression: ad_cotangents isa Tuple
  Problem: The pullback must return a Tuple (∂self, ∂args...)
   Evaluated: [5.351804219822569 0.0 0.0; 0.0 0.7152307254641339 0.0; 0.0 0.0 0.27395514616548344] isa Tuple
Stacktrace:
  [1] #invokelatest#2
    @ ./essentials.jl:708 [inlined]
  [2] invokelatest(::Any)
    @ Base ./essentials.jl:706
  [3] evaluate_call_recurse!(recurse::Any, frame::Frame, call_expr::Expr; enter_generated::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:229
  [4] evaluate_call_recurse!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:201 [inlined]
  [5] eval_rhs(recurse::Any, frame::Frame, node::Expr)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:400
  [6] step_expr!(recurse::Any, frame::Frame, node::Any, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:561
  [7] step_expr!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:611
  [8] finish!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:14
  [9] finish_and_return!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:30 [inlined]
 [10] evaluate_call_recurse!(recurse::Any, frame::Frame, call_expr::Expr; enter_generated::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:240
--- the last 7 lines are repeated 5 more times ---
 [46] evaluate_call_recurse!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:201 [inlined]
 [47] eval_rhs(recurse::Any, frame::Frame, node::Expr)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:400
 [48] step_expr!(recurse::Any, frame::Frame, node::Any, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:561
 [49] step_expr!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:611
 [50] finish!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:14
 [51] finish_and_return!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:30 [inlined]
 [52] finish_and_return! (repeats 2 times)
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:34 [inlined]
 [53] top-level scope
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/construct.jl:781
 [54] include
    @ ./Base.jl:387 [inlined]
 [55] macro expansion
    @ ./timing.jl:210 [inlined]
 [56] include_test(path::String)
    @ Main ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:27
 [57] macro expansion
    @ ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:63 [inlined]
 [58] macro expansion
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
--- the last 2 lines are repeated 1 more time ---
 [61] macro expansion
    @ ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:59 [inlined]
 [62] macro expansion
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
test_rrule: sqrt on Diagonal{Float64, Vector{Float64}}: Test Failed at /Users/thomaswright/.julia/packages/ChainRulesTestUtils/Rzheq/src/testers.jl:215
  Expression: length(ad_cotangents) == length(primals)
  Problem: The pullback should return 1 cotangent for the primal and each primal input.
   Evaluated: 9 == 2
Stacktrace:
  [1] #invokelatest#2
    @ ./essentials.jl:708 [inlined]
  [2] invokelatest(::Any)
    @ Base ./essentials.jl:706
  [3] evaluate_call_recurse!(recurse::Any, frame::Frame, call_expr::Expr; enter_generated::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:229
  [4] evaluate_call_recurse!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:201 [inlined]
  [5] eval_rhs(recurse::Any, frame::Frame, node::Expr)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:400
  [6] step_expr!(recurse::Any, frame::Frame, node::Any, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:561
  [7] step_expr!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:611
  [8] finish!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:14
  [9] finish_and_return!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:30 [inlined]
 [10] evaluate_call_recurse!(recurse::Any, frame::Frame, call_expr::Expr; enter_generated::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:240
--- the last 7 lines are repeated 5 more times ---
 [46] evaluate_call_recurse!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:201 [inlined]
 [47] eval_rhs(recurse::Any, frame::Frame, node::Expr)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:400
 [48] step_expr!(recurse::Any, frame::Frame, node::Any, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:561
 [49] step_expr!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/interpret.jl:611
 [50] finish!(recurse::Any, frame::Frame, istoplevel::Bool)
    @ JuliaInterpreter ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:14
 [51] finish_and_return!
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:30 [inlined]
 [52] finish_and_return! (repeats 2 times)
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/commands.jl:34 [inlined]
 [53] top-level scope
    @ ~/.julia/packages/JuliaInterpreter/4B89D/src/construct.jl:781
 [54] include
    @ ./Base.jl:387 [inlined]
 [55] macro expansion
    @ ./timing.jl:210 [inlined]
 [56] include_test(path::String)
    @ Main ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:27
 [57] macro expansion
    @ ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:63 [inlined]
 [58] macro expansion
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
--- the last 2 lines are repeated 1 more time ---
 [61] macro expansion
    @ ~/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:59 [inlined]
 [62] macro expansion
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
test_rrule: sqrt on Diagonal{Float64, Vector{Float64}}: Error During Test at /Users/thomaswright/.julia/packages/ChainRulesTestUtils/Rzheq/src/testers.jl:191
  Got exception outside of a @test

  1.239840 seconds (3.63 M allocations: 298.411 MiB, 7.67% gc time, 27.89% compilation time)

Test Summary:                                                    | Pass  Fail  Error  Total
ChainRules                                                       |  641     2      2    645
  test_helpers.jl                                                |    8                   8
  rulesets                                                       |  633     2      2    637
    LinearAlgebra                                                |  633     2      2    637
      Structured Matrices                                        |  633     2      2    637
        / and \ on Square Matrixes                               |   81                  81
        Diagonal                                                 |   16                  16
        dot(x, ::Diagonal, y)                                    |   11                  11
        ::Diagonal * ::AbstractVector                            |    9                   9
        diag                                                     |   52                  52
        diagm                                                    |   24                  24
        sqrt(::Diagonal)                                         |    2     2      2      6
          test_rrule: sqrt on Diagonal{Float64, Vector{Float64}} |    1            1      2
          test_rrule: sqrt on Diagonal{Float64, Vector{Float64}} |    1     2      1      4
        Adjoint, Float64                                         |   30                  30
        Adjoint, ComplexF64                                      |   30                  30
        adjoint, Float64                                         |   30                  30
        adjoint, ComplexF64                                      |   30                  30
        Transpose, Float64                                       |   30                  30
        Transpose, ComplexF64                                    |   30                  30
        transpose, Float64                                       |   30                  30
        transpose, ComplexF64                                    |   30                  30
        UpperTriangular                                          |    7                   7
        LowerTriangular                                          |    7                   7
        triu                                                     |   47                  47
        tril                                                     |   47                  47
        det and logdet Diagonal                                  |   30                  30
        det and logdet UpperTriangular                           |   30                  30
        det and logdet LowerTriangular                           |   30                  30
ERROR: LoadError: Some tests did not pass: 641 passed, 2 failed, 2 errored, 0 broken.
in expression starting at /Users/thomaswright/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:34
```

Reply:
I think it may be my fault, that it needs the self-grad: `return (NoTangent, Diagonal(Δ_diag ./ (2 .* y.diag)))`

Reply:
Should we implement the version we know works for now and make an issue regarding improving its allocations?

Reply:
yeah let's do that so that you can finish your project

Reply:
> I think it may be my fault, that it needs the self-grad: `return (NoTangent(), Diagonal(Δ_diag ./ (2 .* y.diag)))`

This got us closer but the error messages (or lack there off) is now an issue:

```
Testing ChainRules.jl
Testing test_helpers.jl:
  0.155204 seconds (229.46 k allocations: 12.906 MiB, 85.97% compilation time)

Testing rulesets/LinearAlgebra/structured.jl:
test_rrule: sqrt on Diagonal{Float64, Vector{Float64}}: Error During Test at /Users/thomaswright/.julia/packages/ChainRulesTestUtils/Rzheq/src/testers.jl:191
  Got exception outside of a @test

test_rrule: sqrt on Diagonal{Float64, Vector{Float64}}: Error During Test at /Users/thomaswright/.julia/packages/ChainRulesTestUtils/Rzheq/src/testers.jl:191
  Got exception outside of a @test

  1.067402 seconds (3.56 M allocations: 294.572 MiB, 7.78% gc time, 30.91% compilation time)

Test Summary:                                                    | Pass  Error  Total
ChainRules                                                       |  643      2    645
  test_helpers.jl                                                |    8             8
  rulesets                                                       |  635      2    637
    LinearAlgebra                                                |  635      2    637
      Structured Matrices                                        |  635      2    637
        / and \ on Square Matrixes                               |   81            81
        Diagonal                                                 |   16            16
        dot(x, ::Diagonal, y)                                    |   11            11
        ::Diagonal * ::AbstractVector                            |    9             9
        diag                                                     |   52            52
        diagm                                                    |   24            24
        sqrt(::Diagonal)                                         |    4      2      6
          test_rrule: sqrt on Diagonal{Float64, Vector{Float64}} |    1      1      2
          test_rrule: sqrt on Diagonal{Float64, Vector{Float64}} |    3      1      4
        Adjoint, Float64                                         |   30            30
        Adjoint, ComplexF64                                      |   30            30
        adjoint, Float64                                         |   30            30
        adjoint, ComplexF64                                      |   30            30
        Transpose, Float64                                       |   30            30
        Transpose, ComplexF64                                    |   30            30
        transpose, Float64                                       |   30            30
        transpose, ComplexF64                                    |   30            30
        UpperTriangular                                          |    7             7
        LowerTriangular                                          |    7             7
        triu                                                     |   47            47
        tril                                                     |   47            47
        det and logdet Diagonal                                  |   30            30
        det and logdet UpperTriangular                           |   30            30
        det and logdet LowerTriangular                           |   30            30
ERROR: LoadError: Some tests did not pass: 643 passed, 0 failed, 2 errored, 0 broken.
in expression starting at /Users/thomaswright/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:34
```

Reply:
> Should we implement the version we know works for now and make an issue regarding improving its allocations?
> > yeah let's do that so that you can finish your project

@mcabbott Are you okay with this? Happy to see if we can get this working right away but if it is not sorted by tonight, I would prefer merging the working solution by the end of the evening and making an issue to improve allocations with your example linked :)

Reply:
actually sorry, maybe try what Michael said, it seems to work for me (for the two tests that are on your remote branch)

Reply:
My vote is usually to try to get things good the first time.

If testing locally, BTW, commenting out pretty much all of `runtests.jl` (but not committing this) will let you see if things work in a minute instead of an hour.

Reply:
I've pushed the changes, the tests seem to work for me, one other thing was just misspelling `diag` as `daig`, is that what shows up in your error messages?

Reply:
It is still giving me an error for the second test here:
```
test_rrule(sqrt, Diagonal(rand(3)))
test_rrule(sqrt, Diagonal(rand(3)); output_tangent=Diagonal(rand(3)) + zeros(3,3))
```

Reply:
that test works for me - what's the error?

Reply:
Does it work using `rand(3)` or the hardcoded `Diagonal([1.0,2,3])` ? Thats what breaks it for me.

```
Testing ChainRules.jl
Testing test_helpers.jl:
  0.175648 seconds (229.30 k allocations: 12.900 MiB, 88.39% compilation time)

Testing rulesets/LinearAlgebra/structured.jl:
test_rrule: sqrt on Diagonal{Float64, Vector{Float64}}: Error During Test at /Users/thomaswright/.julia/packages/ChainRulesTestUtils/Rzheq/src/testers.jl:191
  Got exception outside of a @test

  1.117011 seconds (3.57 M allocations: 294.806 MiB, 8.63% gc time, 30.17% compilation time)

Test Summary:                                                    | Pass  Error  Total
ChainRules                                                       |  649      1    650
  test_helpers.jl                                                |    8             8
  rulesets                                                       |  641      1    642
    LinearAlgebra                                                |  641      1    642
      Structured Matrices                                        |  641      1    642
        / and \ on Square Matrixes                               |   81            81
        Diagonal                                                 |   16            16
        dot(x, ::Diagonal, y)                                    |   11            11
        ::Diagonal * ::AbstractVector                            |    9             9
        diag                                                     |   52            52
        diagm                                                    |   24            24
        sqrt(::Diagonal)                                         |   10      1     11
          test_rrule: sqrt on Diagonal{Float64, Vector{Float64}} |    7             7
          test_rrule: sqrt on Diagonal{Float64, Vector{Float64}} |    3      1      4
        Adjoint, Float64                                         |   30            30
        Adjoint, ComplexF64                                      |   30            30
        adjoint, Float64                                         |   30            30
        adjoint, ComplexF64                                      |   30            30
        Transpose, Float64                                       |   30            30
        Transpose, ComplexF64                                    |   30            30
        transpose, Float64                                       |   30            30
        transpose, ComplexF64                                    |   30            30
        UpperTriangular                                          |    7             7
        LowerTriangular                                          |    7             7
        triu                                                     |   47            47
        tril                                                     |   47            47
        det and logdet Diagonal                                  |   30            30
        det and logdet UpperTriangular                           |   30            30
        det and logdet LowerTriangular                           |   30            30
ERROR: LoadError: Some tests did not pass: 649 passed, 0 failed, 1 errored, 0 broken.
in expression starting at /Users/thomaswright/Documents/Dev_Folder/ChainRules.jl/test/runtests.jl:34
```

Reply:
I am copy-pasting the second example
```julia
test_rrule(sqrt, Diagonal(rand(3)); output_tangent=Diagonal(rand(3)) + zeros(3,3))
```

have you pulled all the changes? also weird that it is not showing what the exception is

Reply:
with your changes I still get this the error with these tests:

```
@testset "sqrt(::Diagonal)" begin
        test_rrule(sqrt, Diagonal(rand(3)))
        test_rrule(sqrt, Diagonal(rand(3)); output_tangent=Diagonal(rand(3)) + zeros(3,3))
    end
```